In [27]:
!pip install pyspark

In [28]:
from pyspark import SparkConf,SparkContext
import collections
#milyonlarca makinaya bağlanıp tüm veriyi getirir

In [35]:
sc=SparkContext()
rdd=sc.parallelize([4,5,2,2])

sq=rdd.map(lambda x:x*x)

print(sq.collect())

sc.stop()

#sc.stop u her satırda yazmamız lazım
#bunu demezsek sürekli bilgi almaya devam eder spark diğer satırda çalışmaz

[16, 25, 4, 4]


## Ratings Histogram¶

We will run a simple Spark script using Python and analyze the 100,000 movie ratings

In [36]:
from pyspark import SparkConf, SparkContext
import collections

#Sparkın localde çalışması ve projemize isim veriyoruz
conf=SparkConf().setMaster("local").setAppName("isim")
sc=SparkContext(conf=conf)

lines=sc.textFile("u.data")
ratings = lines.map(lambda x: x.split()[2])
result= ratings.countByValue()
sortedResults= collections.OrderedDict(sorted(result.items()))
for key, value in sortedResults.items():
    print("%s %i" % (key, value))
    
sc.stop()


1 6111
2 11370
3 27145
4 34174
5 21203


local dediğimiz zaman kendi makinamızı kasdediyoruz.
sc.textfile ile satır satır okur,map ile tüm satırlarda işlemleri yapmış oluyoruz
1 yıldızdan rating 6111 defa tekrar etmiş
collections ile farklı yerlerden bilgi alıyoruz 
stop yazmazsan dışarıda ayrıca stop yazman gerekiyor

spark ta satır satır okuyoruz virgülle ayırıyoruz csv de kendisi okuyordu
yaşı ve o yaştaki arkadaş sayısını geri döndürür
reduce ile birleri okuyor
kaç tane 33 yaşında olan rakadaşım var diyor ve onların ortalamasını alıyor

In [52]:
#sc.stop() # mültiple hatası olursa bunu ayrı bir satırda tekrar çalıştır

In [41]:
from google.colab import files
files.upload()

Saving 1800.csv to 1800.csv


{'1800.csv': b'ITE00100554,18000101,TMAX,-75,,,E,\nITE00100554,18000101,TMIN,-148,,,E,\nGM000010962,18000101,PRCP,0,,,E,\nEZE00100082,18000101,TMAX,-86,,,E,\nEZE00100082,18000101,TMIN,-135,,,E,\nITE00100554,18000102,TMAX,-60,,I,E,\nITE00100554,18000102,TMIN,-125,,,E,\nGM000010962,18000102,PRCP,0,,,E,\nEZE00100082,18000102,TMAX,-44,,,E,\nEZE00100082,18000102,TMIN,-130,,,E,\nITE00100554,18000103,TMAX,-23,,,E,\nITE00100554,18000103,TMIN,-46,,I,E,\nGM000010962,18000103,PRCP,4,,,E,\nEZE00100082,18000103,TMAX,-10,,,E,\nEZE00100082,18000103,TMIN,-73,,,E,\nITE00100554,18000104,TMAX,0,,,E,\nITE00100554,18000104,TMIN,-13,,,E,\nGM000010962,18000104,PRCP,0,,,E,\nEZE00100082,18000104,TMAX,-55,,,E,\nEZE00100082,18000104,TMIN,-74,,,E,\nITE00100554,18000105,TMAX,10,,,E,\nITE00100554,18000105,TMIN,-6,,,E,\nGM000010962,18000105,PRCP,0,,,E,\nEZE00100082,18000105,TMAX,-40,,,E,\nEZE00100082,18000105,TMIN,-58,,,E,\nITE00100554,18000106,TMAX,13,,,E,\nITE00100554,18000106,TMIN,13,,,E,\nGM000010962,18000106,PR

## the Average Friends by Age

In [ ]:
from pyspark import SparkConf, SparkContext

conf = SparkConf().setMaster("local").setAppName("FriendsByAge")
sc = SparkContext(conf = conf)

def parseLine(line):
    fields = line.split(',')
    age = int(fields[2])
    numFriends = int(fields[3])
    return (age, numFriends)

lines = sc.textFile("fakefriends.csv")
rdd = lines.map(parseLine)
totalsByAge = rdd.mapValues(lambda x: (x, 1)).reduceByKey(lambda x, y: (x[0] + y[0], x[1] + y[1]))
averagesByAge = totalsByAge.mapValues(lambda x: int(x[0] / x[1]))
results = averagesByAge.collect()
for result in results:
    print(result)

sc.stop()

Mesela 246 tane 23 yaşında arkadaşım varmış,343 tane 18 yaşında arkadaşım varmış.
App a friends by age diye isim veriyoruz,2.sütun yaşı,3.sütun arkadaş sayısını veriyor,scçtextfile ile csv okuyoruz
map ile fonksiyona gönderiyoruz fonksiyonda split ile , e göre ayırıyoruz

## Filtering RDDs and the Minimum Temperature by Location

TMIN diye bir satır buluyor ve oradan min temperature u bulup geliyor.Burada yaptığı şey 1800 kelime(dökümandan) içinden min temperature u bulmak

In [42]:
from pyspark import SparkConf, SparkContext

conf = SparkConf().setMaster("local").setAppName("MinTemperatures")
sc = SparkContext(conf = conf)

def parseLine(line):
    fields = line.split(',')
    stationID = fields[0]
    entryType = fields[2]
    temperature = float(fields[3]) * 0.1 * (9.0 / 5.0) + 32.0
    return (stationID, entryType, temperature)

lines = sc.textFile("1800.csv")
parsedLines = lines.map(parseLine)
minTemps = parsedLines.filter(lambda x: "TMIN" in x[1])
stationTemps = minTemps.map(lambda x: (x[0], x[2]))
minTemps = stationTemps.reduceByKey(lambda x, y: min(x,y))
results = minTemps.collect();

for result in results:
    print(result[0] + "\t{:.2f}F".format(result[1]))
    
sc.stop()

ITE00100554	5.36F
EZE00100082	7.70F


temperature = float(fields[3]) * 0.1 * (9.0 / 5.0) + 32.0 ile sıcaklık F a çevrilir.
Filter ile bulduğu sıcaklığın en küçüğünü alıyor.Filtreliyor.

## the Maximum Temperature by Location

In [43]:
conf = SparkConf().setMaster("local").setAppName("MaxTemperatures")
sc = SparkContext(conf = conf)

def parseLine(line):
    fields = line.split(',')
    stationID = fields[0]
    entryType = fields[2]
    temperature = float(fields[3]) * 0.1 * (9.0 / 5.0) + 32.0
    return (stationID, entryType, temperature)

lines = sc.textFile("1800.csv")
parsedLines = lines.map(parseLine)
minTemps = parsedLines.filter(lambda x: "TMAX" in x[1])
stationTemps = minTemps.map(lambda x: (x[0], x[2]))
minTemps = stationTemps.reduceByKey(lambda x, y: max(x,y))
results = minTemps.collect();

for result in results:
    print(result[0] + "\t{:.2f}F".format(result[1]))
    
sc.stop()

ITE00100554	90.14F
EZE00100082	90.14F


## Counting Word Occurrences Using flatmap()

Bir kitapta bir kelime kaç kere geçiyor onu bulmak istiyoruz

In [ ]:
from pyspark import SparkConf, SparkContext

conf = SparkConf().setMaster("local").setAppName("WordCount")
sc = SparkContext(conf = conf)

input = sc.textFile("book.txt")
words = input.flatMap(lambda x: x.split())
wordCounts = words.countByValue()

for word, count in wordCounts.items():
    cleanWord = word.encode('ascii', 'ignore')
    if (cleanWord):
        print(cleanWord.decode() + " " + str(count))
sc.stop()

## Improving the Word Count Script with Regular Expressions

Anlamsız kelimeleri anlamlı hale getiriyoruz,sadece harfleri alıyoruz rakamları kaldırıyoruz.

In [ ]:
from pyspark import SparkConf, SparkContext

conf = SparkConf().setMaster("local").setAppName("WordCount")
sc = SparkContext(conf = conf)

input = sc.textFile("book.txt")
words = input.flatMap(lambda x: x.split())
wordCounts = words.countByValue()

for word, count in wordCounts.items():
    cleanWord = word.encode('ascii', 'ignore')
    if (cleanWord):
        print(cleanWord.decode() + " " + str(count))
sc.stop()
import re
from pyspark import SparkConf, SparkContext

def normalizeWords(text):
    return re.compile(r'\W+', re.UNICODE).split(text.lower())

conf = SparkConf().setMaster("local").setAppName("WordCount")
sc = SparkContext(conf = conf)

input = sc.textFile("book.txt")
words = input.flatMap(normalizeWords)
wordCounts = words.countByValue()

for word, count in wordCounts.items():
    cleanWord = word.encode('ascii', 'ignore')
    if (cleanWord):
        print(cleanWord.decode() + " " + str(count))

sc.stop()

#Find the Total Amount Spent by Customer

Müşterilerin ne kadar fiyat harcadıklarını gösterir

In [ ]:
from pyspark import SparkConf,SparkContext

conf=SparkConf().setMaster("local").setAppName("SpendByCustomer")
sc=SparkContext(conf=conf)

def extractCustomerPricePairs(line):
  fields=line.split(",")
  return (int(fields[0]), float(fields[2]))

input=sc.textFile("customer-orders.csv")
mappedInput=input.map(extractCustomerPricePairs)
totalByCustomer=mappedInput.reduceByKey(lambda x, y:x+y)

results=totalByCustomer.collect();
for result in results:
  print(result)

sc.stop()

## Sort By

Müşterilerin harcamalarını büyükten küçüğe sıralar flipped sort by key false true yazmaya göre büyük küçük sıralama durumu değişir.

In [ ]:
from pyspark import SparkConf, SparkContext

conf = SparkConf().setMaster("local").setAppName("SpendByCustomerSorted")
sc = SparkContext(conf = conf)

def extractCustomerPricePairs(line):
    fields = line.split(',')
    return (int(fields[0]), float(fields[2]))

input = sc.textFile("customer-orders.csv")
mappedInput = input.map(extractCustomerPricePairs)
totalByCustomer = mappedInput.reduceByKey(lambda x, y: x + y)

#Changed for Python 3 compatibility:
#flipped = totalByCustomer.map(lambda (x,y):(y,x))
flipped = totalByCustomer.map(lambda x: (x[1], x[0]))

totalByCustomerSorted = flipped.sortByKey()

results = totalByCustomerSorted.collect();
for result in results:
    print(result)

sc.stop()


##En çok reyting almış filmleri sıralıyoruz

In [ ]:
from pyspark import SparkConf, SparkContext

conf = SparkConf().setMaster("local").setAppName("PopularMovies")
sc = SparkContext(conf = conf)

lines = sc.textFile("u.data")
movies = lines.map(lambda x: (int(x.split()[1]), 1))
movieCounts = movies.reduceByKey(lambda x, y: x + y)

flipped = movieCounts.map( lambda xy : (xy[1], xy[0]) )
sortedMovies = flipped.sortByKey(False) #(False) Descending order

results = sortedMovies.collect()

for result in results:
    print(result)

sc.stop()


## En çok reytingli filmi, filmin ismini de ekleyerek alıyoruz

In [ ]:
from pyspark import SparkConf, SparkContext

def loadMovieNames():
    movieNames = {}
    with open("u.item", encoding = "ISO-8859-1") as f:
        for line in f:
            fields = line.split('|')
            movieNames[int(fields[0])] = fields[1]
    return movieNames

conf = SparkConf().setMaster("local").setAppName("PopularMovies")
sc = SparkContext(conf = conf)

nameDict = sc.broadcast(loadMovieNames())

lines = sc.textFile("u.data")
movies = lines.map(lambda x: (int(x.split()[1]), 1))
movieCounts = movies.reduceByKey(lambda x, y: x + y)

flipped = movieCounts.map(lambda xy : (xy[1], xy[0]))
sortedMovies = flipped.sortByKey(False)

sortedMoviesWithNames = sortedMovies.map(lambda countMovie : (nameDict.value[countMovie[1]], countMovie[0]))

results = sortedMoviesWithNames.collect()

for result in results:
    print(result)
    
sc.stop()


En popüler süper kahramanı buluyor verilerin içinden

In [53]:
from pyspark import SparkConf, SparkContext

conf = SparkConf().setMaster("local").setAppName("PopularHero")
sc = SparkContext(conf = conf)

def countCoOccurences(line):
    elements = line.split()
    return (int(elements[0]), len(elements) - 1)

def parseNames(line):
    fields = line.split('\"')
    return (int(fields[0]), fields[1].encode("utf8"))

names = sc.textFile("Marvel-names.txt")
namesRdd = names.map(parseNames)

lines = sc.textFile("Marvel-graph.txt")

pairings = lines.map(countCoOccurences)
totalFriendsByCharacter = pairings.reduceByKey(lambda x, y : x + y)
flipped = totalFriendsByCharacter.map(lambda xy : (xy[1], xy[0]))

mostPopular = flipped.max()

mostPopularName = namesRdd.lookup(mostPopular[1])[0]

print(str(mostPopularName) + " is the most popular superhero, with " + str(mostPopular[0]) + \
      " co-appearances.")

sc.stop()


b'CAPTAIN AMERICA' is the most popular superhero, with 1933 co-appearances.


## Introducing SparkSQL¶
We'll cover the concepts of SparkSQL, DataFrames, and DataSets, and why they are so important in Spark 2.0 and above.

## A real example, revisiting our fake social network data and analyzing it with DataFrames through a SparkSession object

Executing SQL Commands and SQL-Style Functions on a DataFrame

In [ ]:
from pyspark.sql import SparkSession
from pyspark.sql import Row

import collections

# Create a SparkSession (Note, the config section is only for Windows!)
spark = SparkSession.builder.config("spark.sql.warehouse.dir").appName("SparkSQL").getOrCreate()

def mapper(line):
    fields = line.split(',')
    return Row(ID=int(fields[0]), name=str(fields[1].encode("utf-8")), age=int(fields[2]), numFriends=int(fields[3]))

lines = spark.sparkContext.textFile("spark/fakefriends.csv")
people = lines.map(mapper)

# Infer the schema, and register the DataFrame as a table.
schemaPeople = spark.createDataFrame(people).cache()
schemaPeople.createOrReplaceTempView("people")

# SQL can be run over DataFrames that have been registered as a table.
teenagers = spark.sql("SELECT * FROM people WHERE age >= 13 AND age <= 19")

# The results of SQL queries are RDDs and support all the normal RDD operations.
for teen in teenagers.collect():
  print(teen)

# We can also use functions instead of SQL queries:
schemaPeople.groupBy("age").count().orderBy("age").show()

spark.stop()


# Using DataFrames Instead of RDDs

In [ ]:
from pyspark.sql import SparkSession
from pyspark.sql import Row
from pyspark.sql import functions

def loadMovieNames():
    movieNames = {}
    with open("spark/ml-100k/u.item") as f:
        for line in f:
            fields = line.split('|')
            movieNames[int(fields[0])] = fields[1]
    return movieNames

# Create a SparkSession (the config bit is only for Windows!)
spark = SparkSession.builder.config("spark.sql.warehouse.dir", "file:///C:/temp").appName("PopularMovies").getOrCreate()

# Load up our movie ID -> name dictionary
nameDict = loadMovieNames()

# Get the raw data
lines = spark.sparkContext.textFile("spark/ml-100k/u.data")
# Convert it to a RDD of Row objects
movies = lines.map(lambda x: Row(movieID =int(x.split()[1])))
# Convert that to a DataFrame
movieDataset = spark.createDataFrame(movies)

# Some SQL-style magic to sort all movies by popularity in one line!
topMovieIDs = movieDataset.groupBy("movieID").count().orderBy("count", ascending=False).cache()

# Show the results at this point:

#|movieID|count|
#+-------+-----+
#|     50|  584|
#|    258|  509|
#|    100|  508|

topMovieIDs.show()

# Grab the top 10
top10 = topMovieIDs.take(10)

# Print the results
print("\n")
for result in top10:
    # Each row has movieID, count as above.
    print("%s: %d" % (nameDict[result[0]], result[1]))

# Stop the session
spark.stop()

# Introducing MLLIB¶

We'll briefly cover the capabilities of Spark's MLLib machine learning library, and how it can help you solve data mining, machine learning, and statistical problems you may encounter. We'll go into more depth on MLLib's Alternating Least Squares (ALS) recommendation engine, and how we can use it to produce movie recommendations with the MovieLens dataset.

In [ ]:
import sys
from pyspark import SparkConf, SparkContext
from pyspark.mllib.recommendation import ALS, Rating

def loadMovieNames():
    movieNames = {}
    with open("spark/ml-100k/u.item") as f:
        for line in f:
            fields = line.split('|')
            movieNames[int(fields[0])] = fields[1] #.decode('ascii', 'ignore')
    return movieNames

conf = SparkConf().setMaster("local[*]").setAppName("MovieRecommendationsALS")
sc = SparkContext(conf = conf)
sc.setCheckpointDir('checkpoint')

print("\nLoading movie names...")
nameDict = loadMovieNames()

data = sc.textFile("spark/ml-100k/u.data")

ratings = data.map(lambda l: l.split()).map(lambda l: Rating(int(l[0]), int(l[1]), float(l[2]))).cache()

# Build the recommendation model using Alternating Least Squares
print("\nTraining recommendation model...")
rank = 10
# Lowered numIterations to ensure it works on lower-end systems
numIterations = 6
model = ALS.train(ratings, rank, numIterations)

userID = int(sys.argv[1])

print("\nRatings for user ID " + str(userID) + ":")
userRatings = ratings.filter(lambda l: l[0] ==0)
for rating in userRatings.collect():
    print (nameDict[int(rating[1])] + ": " + str(rating[2]))

print("\nTop 10 recommendations:")
recommendations = model.recommendProducts(userID, 10)
for recommendation in recommendations:
    print (nameDict[int(recommendation[1])] + " score " + str(recommendation[2]))


sc.stop()

# Linear Regression with Spark

In [ ]:
from __future__ import print_function

from pyspark.ml.regression import LinearRegression

from pyspark.sql import SparkSession
from pyspark.ml.linalg import Vectors

# Create a SparkSession (Note, the config section is only for Windows!)
spark = SparkSession.builder.config("spark.sql.warehouse.dir", "file:///C:/temp").appName("LinearRegression").getOrCreate()

# Load up our data and convert it to the format MLLib expects.
inputLines = spark.sparkContext.textFile("spark/regression.txt")
data = inputLines.map(lambda x: x.split(",")).map(lambda x: (float(x[0]), Vectors.dense(float(x[1]))))

# Convert this RDD to a DataFrame
colNames = ["label", "features"]
df = data.toDF(colNames)

# Note, there are lots of cases where you can avoid going from an RDD to a DataFrame.
# Perhaps you're importing data from a real database. Or you are using structured streaming
# to get your data.

# Let's split our data into training data and testing data
trainTest = df.randomSplit([0.5, 0.5])
trainingDF = trainTest[0]
testDF = trainTest[1]

# Now create our linear regression model
lir = LinearRegression(maxIter=10, regParam=0.3, elasticNetParam=0.8)

# Train the model using our training data
model = lir.fit(trainingDF)

# Now see if we can predict values in our test data.
# Generate predictions using our linear regression model for all features in our
# test dataframe:
fullPredictions = model.transform(testDF).cache()

# Extract the predictions and the "known" correct labels.
predictions = fullPredictions.select("prediction").rdd.map(lambda x: x[0])
labels = fullPredictions.select("label").rdd.map(lambda x: x[0])

# Zip them together
predictionAndLabel = predictions.zip(labels).collect()

# Print out the predicted and actual values for each point
for prediction in predictionAndLabel:
  print(prediction)


# Stop the session
spark.stop()